src/error_diffusion.jl

```diff
@@ -2,52 +2,41 @@ function error_diffusion(
     img::AbstractMatrix{<:Gray}, stencil::OffsetMatrix; to_linear=false
 )::BitMatrix
     # Change from normalized intensities to Float as error will get added!
-    _img = floattype(eltype(img)).(img)
-
-    # Optionally cast to linear colorspace
-    to_linear && srgb2linear!(_img)
+    FT = floattype(eltype(img))
+    if to_linear
+        _img = @. FT(srgb2linear(img))
+    else
+        _img = FT.(img)
+    end
+    stencil = FT.(stencil) # eagerly promote to the same type to make loop run faster
 
-    h, w = size(_img)
-    dither = BitArray(undef, h, w) # initialized to zero
+    dither = BitArray(undef, size(_img)...) # initialized to zero
 
     # Get OffsetMatrix indices
-    drs, dcs = indices(stencil)
-
-    for r in 1:h
-        for c in 1:w
-            px = _img[r, c]
-
-            # Round to closest color
-            px >= 0.5 ? (rnd = 1) : (rnd = 0)
-
-            # Apply pixel to dither
-            dither[r, c] = rnd
-
-            # Diffuse "error" to neighborhood in stencil
-            err = px - rnd
-            for dr in drs
-                for dc in dcs
-                    if (r + dr > 0) && (r + dr <= h) && (c + dc > 0) && (c + dc <= w)
-                        _img[r + dr, c + dc] += err * stencil[dr, dc]
-                    end
-                end
-            end
+    O = CartesianIndices(stencil)
+    R = CartesianIndices(_img)
+    i_first, i_last = first(R), last(R)
+    for p in CartesianIndices(_img)
```

Comment:
The error diffusion algorithm assumes iteration along rows. This leads to a different output as `CartesianIndices` iterate over columns first.

Reply:
> The error diffusion algorithm assumes iteration along rows.

I think this assumption is based on that other languages are row-major order. In Julia, arrays are column-major order so indexing along columns is the fastest.

If we interpret the Julia array as a C pointer, then it is faithfully "along rows" in the C/Python world.

Reply:
Yes, but this would also require all error diffusion "stencils" to be changed. All stencils are asymmetrical, e.g. the Matrix used for `floyd_steinberg_diffusion`:
```julia
const FS_STENCIL = OffsetArray([0 0 7; 3 5 1]//16, 0:1, -1:1)
```
These describe the neighbourhood into which the error of the color discretisation is "diffused". The direction in which this error is diffused first has a big influence on the output:
* `master`:
<img width="715" alt="dither_master" src="https://user-images.githubusercontent.com/20258504/117293923-50cc4080-ae72-11eb-8adc-e70c4fc719de.png">

* this PR:
<img width="715" alt="dither_PR" src="https://user-images.githubusercontent.com/20258504/117293737-1367b300-ae72-11eb-956f-8a1777f21b8c.png">

We could "transpose" the stencils, but the output would still look different from a row-wise application on the same image.

See e.g. [this description of the Floyd–Steinberg algorithm](https://en.wikipedia.org/wiki/Floyd–Steinberg_dithering):
> The algorithm scans the image from left to right, top to bottom, quantizing pixel values one by one.

(By the way, I fully agree correctness tests are overdue...)

Reply:
In cfed6e8, I've reverted the "fancy" `CartesianIndices` rewrite, and it turns out to be more efficient:

```julia
@btime simple_error_diffusion($img);
# master: 743.018 μs (4 allocations: 232.31 KiB)
# before: 481.427 μs (5 allocations: 232.41 KiB)
# after: 392.075 μs (5 allocations: 232.41 KiB)

@btime simple_error_diffusion($img, to_linear=true);
# master: 1.753 ms (4 allocations: 232.31 KiB)
# before: 562.010 μs (5 allocations: 232.41 KiB)
# after: 496.748 μs (5 allocations: 232.41 KiB)
```